apps/backend/src/app/api/latest/emails/dev-server/chat/route.tsx

```diff
@@ -0,0 +1,196 @@
+import { getEnvVariable } from "@stackframe/stack-shared/dist/utils/env";
+import { StatusError } from "@stackframe/stack-shared/dist/utils/errors";
+import { convertToCoreMessages, generateText, tool } from "ai";
+import { FreestyleSandboxes } from "freestyle-sandboxes";
+import { prismaClient } from "@/prisma-client";
+import { InferType } from "yup";
+import { z } from "zod";
+
+const textContentSchema = yupObject({
+  type: yupString().oneOf(["text"]).defined(),
+  text: yupString().defined(),
+});
+
+const toolCallContentSchema = yupObject({
+  type: yupString().oneOf(["tool-call"]).defined(),
+  toolName: yupString().defined(),
+  toolCallId: yupString().defined(),
+  args: yupMixed().defined(),
+  argsText: yupString().defined(),
+  result: yupMixed().defined(),
+});
+
+const contentSchema = yupArray(yupUnion(textContentSchema, toolCallContentSchema)).defined();
+
+
+export const POST = createSmartRouteHandler({
+  metadata: {
+    summary: "Chat with the dev server",
+    description: "Chat with the dev server to get help with email theme development",
+    tags: ["Emails"],
+  },
+  request: yupObject({
+    auth: yupObject({
+      type: yupString().oneOf(["admin"]).defined(),
+      tenancy: adaptSchema,
+    }),
+    body: yupObject({
+      repo_id: yupString().defined(),
+      messages: yupArray(yupObject({
+        role: yupString().oneOf(["user", "assistant"]).defined(),
+        content: yupString().defined(),
+      })).defined(),
+    }),
+  }),
+  response: yupObject({
+    statusCode: yupNumber().oneOf([200]).defined(),
+    bodyType: yupString().oneOf(["json"]).defined(),
+    body: yupObject({
+      content: contentSchema,
+    }).defined(),
+  }),
+  async handler({ body, auth: { tenancy } }) {
+    const apiKey = getEnvVariable("STACK_FREESTYLE_API_KEY");
+    if (!apiKey) {
+      throw new StatusError(500, "STACK_FREESTYLE_API_KEY is not set");
+    }
+    const userContent = body.messages.at(-1);
+    if (!userContent) {
+      throw new StatusError(400, "No user content");
+    }
+
+    const freestyle = new FreestyleSandboxes({ apiKey });
+    const { fs } = await freestyle.requestDevServer({ repoId: body.repo_id });
+    const currentEmailTheme = await fs.readFile("src/email-theme.tsx");
+    const result = await generateText({
+      model: openai("gpt-4o"),
+      system: DEFAULT_SYSTEM_PROMPT,
+      messages: convertToCoreMessages(body.messages),
+      tools: {
+        createEmailTheme: tool({
+          description: CREATE_EMAIL_THEME_TOOL_DESCRIPTION(currentEmailTheme),
+          parameters: z.object({
+            content: z.string().describe("The content of the email theme"),
+          }),
+          execute: async (args) => {
+            const { fs } = await freestyle.requestDevServer({ repoId: body.repo_id });
+            await fs.writeFile("src/email-theme.tsx", args.content);
```

Comment:
can we wrap this and all other freestyle API calls in a traceSpan (imported from telemetry.tsx)? that way, we can later measure how fast or slow freestyle's API is

(particularly curious about the dev server speed)